docs/proposals/PROPOSAL_B.md

```diff
@@ -0,0 +1,287 @@
+            "vfio",
+            "cpu",
+            "gpu"
+          ]
+        }
+      }
+    },
+    "vfio": {
+      "plugins": {
+        "kernelConfiguration": {
+          "data": {
+            "CONFIG_VFIO": "m",
+            "CONFIG_VFIO_IOMMU_TYPE1": "m",
+            "CONFIG_VFIO_MDEV": "m",
+            "CONFIG_VFIO_MDEV_DEVICE": "m",
+            "CONFIG_VFIO_PCI": "m"
+          }
+        },
+        "kernelModules": {
+          "data": {
+            "vfio": {},
+            "vfio_iommu_type1": {},
+            "vfio-pci": {}
+          }
+        }
+      }
+    },
+    "cpu": {
+      "refs": {
+        "oneOf": [
+          "cpuAmd",
+          "cpuIntel"
+        ]
+      }
+    },
+    "cpuAmd": {
+      "plugins": {
+        "hardware": {
+          "data": {
+            "cpu": ["AuthenticAMD"]
+          }
+        }
+      }
+    },
+    "cpuIntel": {
+      "plugins": {
+        "hardware": {
+          "data": {
+            "cpu": ["GenuineIntel"]
+          }
+        },
+        "kernelCmdline": {
+          "data": {
+            "intel_iommu": "on"
+          }
+        }
+      }
+    },
+    "gpu": {
+      "refs": {
+        "oneOf": [
+          "gpuNVIDIA",
+          "gpuIntel"
+        ]
+      }
+    },
+    "gpuNVIDIA": {
+      "plugins": {
+        "hardware": {
+          "data": {
+            "pci": ["0380:10de"]
+          }
+        },
+        "kernelConfiguration": {
+          "data": {
+            "CONFIG_HOTPLUG_PCI_PCIE": "y",
+            "CONFIG_MODULES": "y",
+            "CONFIG_MODULE_UNLOAD": "y",
+            "CONFIG_PCI_MMCONFIG": "y",
+            "CONFIG_DRM_NOUVEAU": "n"
+          }
+        }
+      }
+    },
+    "gpuIntel": {
+      "plugins": {
+        "hardware": {
+          "data": {
+            "pci": ["0380:8086"]
+          }
+        },
+        "kernelConfiguration": {
+          "data": {
+            "CONFIG_DRM": "y",
+            "CONFIG_DRM_I915": "y",
+            "CONFIG_DRM_I915_USERPTR": "y",
+            "CONFIG_DRM_I915_GVT": "m",
+            "CONFIG_DRM_I915_GVT_KVMGT": "m"
+          }
+        },
+        "kernelModules": {
+          "data": {
+            "kvmgt": {
+              "enable_gvt": "1"
+            }
+          }
+        }
+      }
+    }
+  }
+}
+```
+
+</details>
+
+### Namespace and Plugin
+
+!["Namespace and Plugin](img/B_ns-plugin.png "Namespace and Plugin")
+
+Namespaces and plugins are core functionality to evaluate if a host is compatible with a container.
+They can look up existing components on a host or just serve as a data source to get additional information about specific resources (like CPU features).
+Although in each case a plugin has to return information about success or failure.
+
+The design allows you to build plugins on top of other plugins.
+For instance an Nvidia GPU plugin could include logic of kernelModules and kernelConfiguration to discover if appropriate kernel modules are loaded or disabled.
```

Comment:
This is saying the plugins (binaries) that use other plugins have that relationship represented in two places - in the json graph here, _and_ the plugin. I would argue to remove it from the json graph and allow the plugin to handle the logic. E.g., if it finds the other namespaces that it needs (and some request to do so from a user or running tool) then it can do the check. We are going to run into issues always having to update two things - what happens if the plugin binary changes but a json object hasn't been updated? What happens if the plugin needs to use new namespaces (we have to update the plugin and json instead of just the plugin). There are many things that can go wrong if these two things aren't in sync, so I would put the logic in one place, with the plugin, and allow the plugin to find / use namespaces that it needs.

Reply:
Probably I should describe this point better.

My point was.. developers who write their own plugins could import the code from atomic plugins to avoid logic duplication. For example if I write NVIDIA GPU plugin, I have to check kernel configuration and modules, so I could already import the existing code that implements the logic from other plugins and then build my new binary instead of writing everything again.

It is not about representing the plugins internal logic on a graph. I have to rephrease this point to make that clear.

Reply:
However, you have a good point about plugins versioning. In the plugin spec I should cover the version of the plugin itself.